[Question] Add MDC ping/pong latency test

segabriel opened the pull request:
Hi guys,
It's a lot to just ask, so I've created PR.
I wrote simple (I think so) example with MDC, I used separate embedded MediaDriver on server and client sides. And I applied the following VM options:

```c
 -XX:BiasedLockingStartupDelay=0
 -XX:+UnlockDiagnosticVMOptions
 -XX:GuaranteedSafepointInterval=300000
 -Djava.net.preferIPv4Stack=true
 -Daeron.mtu.length=1k
 -Daeron.socket.so_sndbuf=4k
 -Daeron.socket.so_rcvbuf=4k
 -Daeron.rcv.initial.window.length=4k
 -Dagrona.disable.bounds.checks=true
 -Daeron.term.buffer.sparse.file=false
```

System props:

```c
    $ sysctl net.core.wmem_max
    net.core.wmem_max = 2097152
    $ sysctl net.core.rmem_max
    net.core.rmem_max = 2097152
    $ sysctl net.core.wmem_default
    net.core.wmem_default = 65536
    $ sysctl net.core.rmem_default
    net.core.rmem_default = 65536
```

And here results on my laptop:

```c
       Value     Percentile TotalCount 1/(1-Percentile)

  299630.591 0.000000000000        288           1.00
  309067.775 0.100000000000      43752           1.11
  310116.351 0.200000000000      90160           1.25
  313262.079 0.300000000000     126896           1.43
  317194.239 0.400000000000     169114           1.67
  321126.399 0.500000000000     214355           2.00
  322174.975 0.550000000000     240587           2.22
  322961.407 0.600000000000     259384           2.50
  324009.983 0.650000000000     275984           2.86
  324796.415 0.700000000000     296920           3.33
  333447.167 0.750000000000     320236           4.00
  333971.455 0.775000000000     329324           4.44
  335282.175 0.800000000000     346248           5.00
  335544.319 0.825000000000     360368           5.71
  335544.319 0.850000000000     360368           6.67
  335806.463 0.875000000000     379220           8.00
  335806.463 0.887500000000     379220           8.89
  336068.607 0.900000000000     397264          10.00
  336068.607 0.912500000000     397264          11.43
  336068.607 0.925000000000     397264          13.33
  336068.607 0.937500000000     397264          16.00
  336330.751 0.943750000000     409192          17.78
  336330.751 0.950000000000     409192          20.00
  336330.751 0.956250000000     409192          22.86
  336330.751 0.962500000000     409192          26.67
  336592.895 0.968750000000     417409          32.00
  336592.895 0.971875000000     417409          35.56
  336592.895 0.975000000000     417409          40.00
  336592.895 0.978125000000     417409          45.71
  336592.895 0.981250000000     417409          53.33
  336592.895 0.984375000000     417409          64.00
  336592.895 0.985937500000     417409          71.11
  336592.895 0.987500000000     417409          80.00
  336855.039 0.989062500000     422072          91.43
  336855.039 0.990625000000     422072         106.67
  336855.039 0.992187500000     422072         128.00
  336855.039 0.992968750000     422072         142.22
  336855.039 0.993750000000     422072         160.00
  336855.039 0.994531250000     422072         182.86
  336855.039 0.995312500000     422072         213.33
  336855.039 0.996093750000     422072         256.00
  336855.039 0.996484375000     422072         284.44
  336855.039 0.996875000000     422072         320.00
  336855.039 0.997265625000     422072         365.71
  336855.039 0.997656250000     422072         426.67
  336855.039 0.998046875000     422072         512.00
  336855.039 0.998242187500     422072         568.89
  336855.039 0.998437500000     422072         640.00
  336855.039 0.998632812500     422072         731.43
  336855.039 0.998828125000     422072         853.33
  336855.039 0.999023437500     422072        1024.00
  337117.183 0.999121093750     422480        1137.78
  337117.183 1.000000000000     422480
#[Mean    =   321405.290, StdDeviation   =    10019.345]
#[Max     =   337117.183, Total count    =       422480]
#[Buckets =           32, SubBuckets     =         2048]
```

Results of running `aeron-samples/scripts/embedded-ping-pong`:

```c
       Value     Percentile TotalCount 1/(1-Percentile)

       6.855 0.000000000000          1           1.00
       8.743 0.100000000000    1018220           1.11
       8.831 0.200000000000    2049613           1.25
       8.879 0.300000000000    3084015           1.43
       8.935 0.400000000000    4071445           1.67
       9.759 0.500000000000    5000984           2.00
       9.999 0.550000000000    5525207           2.22
      10.071 0.600000000000    6087520           2.50
      10.119 0.650000000000    6542128           2.86
      10.167 0.700000000000    7029586           3.33
      10.279 0.750000000000    7511240           4.00
      10.935 0.775000000000    7762495           4.44
      11.039 0.800000000000    8015445           5.00
      11.311 0.825000000000    8254007           5.71
      11.967 0.850000000000    8501834           6.67
      12.455 0.875000000000    8753159           8.00
      12.599 0.887500000000    8879780           8.89
      12.919 0.900000000000    9000926          10.00
      13.743 0.912500000000    9127638          11.43
      14.095 0.925000000000    9250687          13.33
      14.823 0.937500000000    9376368          16.00
      15.031 0.943750000000    9438612          17.78
      15.199 0.950000000000    9500599          20.00
      15.359 0.956250000000    9563965          22.86
      15.959 0.962500000000    9625251          26.67
      16.335 0.968750000000    9689312          32.00
      16.463 0.971875000000    9719556          35.56
      17.039 0.975000000000    9750517          40.00
      17.295 0.978125000000    9781929          45.71
      17.503 0.981250000000    9814893          53.33
      17.855 0.984375000000    9843928          64.00
      18.383 0.985937500000    9859620          71.11
      18.735 0.987500000000    9875899          80.00
      19.311 0.989062500000    9890682          91.43
      20.495 0.990625000000    9906269         106.67
      22.271 0.992187500000    9921930         128.00
      23.343 0.992968750000    9929690         142.22
      24.479 0.993750000000    9937537         160.00
      25.759 0.994531250000    9945378         182.86
      27.343 0.995312500000    9953171         213.33
      29.807 0.996093750000    9960939         256.00
      31.535 0.996484375000    9964859         284.44
      33.823 0.996875000000    9968755         320.00
      37.791 0.997265625000    9972668         365.71
      47.871 0.997656250000    9976563         426.67
      60.799 0.998046875000    9980470         512.00
      65.087 0.998242187500    9982425         568.89
      68.031 0.998437500000    9984404         640.00
      72.063 0.998632812500    9986334         731.43
      77.055 0.998828125000    9988290         853.33
      88.063 0.999023437500    9990237        1024.00
      97.535 0.999121093750    9991212        1137.78
     109.439 0.999218750000    9992203        1280.00
     111.103 0.999316406250    9993186        1462.86
     122.047 0.999414062500    9994141        1706.67
     143.103 0.999511718750    9995122        2048.00
     159.103 0.999560546875    9995606        2275.56
     209.535 0.999609375000    9996095        2560.00
     215.167 0.999658203125    9996599        2925.71
     221.823 0.999707031250    9997072        3413.33
     326.143 0.999755859375    9997559        4096.00
     387.583 0.999780273438    9997803        4551.11
     428.543 0.999804687500    9998047        5120.00
     507.135 0.999829101563    9998293        5851.43
     648.703 0.999853515625    9998536        6826.67
     966.143 0.999877929688    9998780        8192.00
    1172.479 0.999890136719    9998902        9102.22
    1583.103 0.999902343750    9999024       10240.00
    1873.919 0.999914550781    9999146       11702.86
    2584.575 0.999926757813    9999268       13653.33
    4012.031 0.999938964844    9999397       16384.00
    4024.319 0.999945068359    9999460       18204.44
    4968.447 0.999951171875    9999513       20480.00
    7327.743 0.999957275391    9999573       23405.71
    7802.879 0.999963378906    9999635       27306.67
    8011.775 0.999969482422    9999699       32768.00
    8032.255 0.999972534180    9999726       36408.89
    8228.863 0.999975585938    9999756       40960.00
    8560.639 0.999978637695    9999787       46811.43
    9183.231 0.999981689453    9999819       54613.33
   10125.311 0.999984741211    9999848       65536.00
   10919.935 0.999986267090    9999863       72817.78
   11829.247 0.999987792969    9999878       81920.00
   12214.271 0.999989318848    9999894       93622.86
   12943.359 0.999990844727    9999909      109226.67
   13811.711 0.999992370605    9999924      131072.00
   14311.423 0.999993133545    9999932      145635.56
   15040.511 0.999993896484    9999939      163840.00
   15572.991 0.999994659424    9999947      187245.71
   16277.503 0.999995422363    9999955      218453.33
   17383.423 0.999996185303    9999963      262144.00
   18104.319 0.999996566772    9999966      291271.11
   18907.135 0.999996948242    9999970      327680.00
   19283.967 0.999997329712    9999974      374491.43
   19775.487 0.999997711182    9999978      436906.67
   20545.535 0.999998092651    9999981      524288.00
   21446.655 0.999998283386    9999983      582542.22
   21479.423 0.999998474121    9999985      655360.00
   21889.023 0.999998664856    9999987      748982.86
   22790.143 0.999998855591    9999989      873813.33
   23674.879 0.999999046326    9999991     1048576.00
   24002.559 0.999999141693    9999992     1165084.44
   24657.919 0.999999237061    9999993     1310720.00
   24674.303 0.999999332428    9999994     1497965.71
   26640.383 0.999999427795    9999995     1747626.67
   27049.983 0.999999523163    9999996     2097152.00
   27049.983 0.999999570847    9999996     2330168.89
   29622.271 0.999999618530    9999997     2621440.00
   29622.271 0.999999666214    9999997     2995931.43
   30212.095 0.999999713898    9999998     3495253.33
   30212.095 0.999999761581    9999998     4194304.00
   30212.095 0.999999785423    9999998     4660337.78
   30277.631 0.999999809265    9999999     5242880.00
   30277.631 0.999999833107    9999999     5991862.86
   30277.631 0.999999856949    9999999     6990506.67
   30277.631 0.999999880791    9999999     8388608.00
   30277.631 0.999999892712    9999999     9320675.55
   30457.855 0.999999904633   10000000    10485760.00
   30457.855 1.000000000000   10000000
#[Mean    =       11.115, StdDeviation   =       80.516]
#[Max     =    30457.855, Total count    =     10000000]
#[Buckets =           24, SubBuckets     =         2048]
```

Also, I launched  `aeron/aeron-samples/scripts/pong.sh` `aeron/aeron-samples/scripts/ping.sh` 
with enabled `-Daeron.sample.embeddedMediaDriver=true` for each ones:

```c
       Value     Percentile TotalCount 1/(1-Percentile)

      18.479 0.000000000000          1           1.00
      30.287 0.100000000000     100143           1.11
      36.543 0.200000000000     200200           1.25
      44.223 0.300000000000     300189           1.43
      52.895 0.400000000000     400495           1.67
      65.855 0.500000000000     500201           2.00
      67.327 0.550000000000     551553           2.22
      69.055 0.600000000000     600981           2.50
      72.319 0.650000000000     650765           2.86
      75.711 0.700000000000     700146           3.33
      81.023 0.750000000000     750320           4.00
      84.479 0.775000000000     775194           4.44
      86.271 0.800000000000     800621           5.00
      88.319 0.825000000000     825075           5.71
      93.247 0.850000000000     850015           6.67
      99.007 0.875000000000     875190           8.00
     100.735 0.887500000000     887974           8.89
     102.207 0.900000000000     900247          10.00
     103.423 0.912500000000     912663          11.43
     104.767 0.925000000000     925184          13.33
     108.031 0.937500000000     937599          16.00
     111.231 0.943750000000     943840          17.78
     115.839 0.950000000000     950023          20.00
     123.007 0.956250000000     956260          22.86
     127.807 0.962500000000     962539          26.67
     136.191 0.968750000000     968854          32.00
     138.367 0.971875000000     971945          35.56
     141.951 0.975000000000     975399          40.00
     142.463 0.978125000000     978478          45.71
     143.615 0.981250000000     981310          53.33
     148.991 0.984375000000     984427          64.00
     153.983 0.985937500000     985938          71.11
     165.247 0.987500000000     987512          80.00
     179.711 0.989062500000     989088          91.43
     184.447 0.990625000000     990641         106.67
     199.935 0.992187500000     992192         128.00
     229.375 0.992968750000     992970         142.22
     309.247 0.993750000000     993751         160.00
     729.599 0.994531250000     994534         182.86
    1787.903 0.995312500000     995313         213.33
    2586.623 0.996093750000     996114         256.00
    2603.007 0.996484375000     996495         284.44
    2619.391 0.996875000000     996909         320.00
    2633.727 0.997265625000     997302         365.71
    2652.159 0.997656250000     997700         426.67
    2666.495 0.998046875000     998057         512.00
    2742.271 0.998242187500     998246         568.89
    3631.103 0.998437500000     998452         640.00
    3657.727 0.998632812500     998637         731.43
    3676.159 0.998828125000     998833         853.33
    3694.591 0.999023437500     999040        1024.00
    3706.879 0.999121093750     999132        1137.78
    3715.071 0.999218750000     999236        1280.00
    3737.599 0.999316406250     999318        1462.86
    4071.423 0.999414062500     999418        1706.67
    4755.455 0.999511718750     999512        2048.00
    4800.511 0.999560546875     999561        2275.56
    4829.183 0.999609375000     999616        2560.00
    4882.431 0.999658203125     999663        2925.71
    5771.263 0.999707031250     999710        3413.33
    5824.511 0.999755859375     999756        4096.00
    5853.183 0.999780273438     999784        4551.11
    5885.951 0.999804687500     999806        5120.00
    5922.815 0.999829101563     999830        5851.43
    6778.879 0.999853515625     999854        6826.67
    6877.183 0.999877929688     999879        8192.00
    6930.431 0.999890136719     999893        9102.22
    7311.359 0.999902343750     999903       10240.00
    8007.679 0.999914550781     999915       11702.86
    8187.903 0.999926757813     999927       13653.33
    8806.399 0.999938964844     999939       16384.00
    8962.047 0.999945068359     999946       18204.44
    9052.159 0.999951171875     999952       20480.00
    9084.927 0.999957275391     999959       23405.71
    9134.079 0.999963378906     999964       27306.67
    9428.991 0.999969482422     999970       32768.00
    9838.591 0.999972534180     999973       36408.89
    9920.511 0.999975585938     999976       40960.00
    9986.047 0.999978637695     999979       46811.43
   10133.503 0.999981689453     999982       54613.33
   10280.959 0.999984741211     999985       65536.00
   10575.871 0.999986267090     999987       72817.78
   10592.255 0.999987792969     999988       81920.00
   10829.823 0.999989318848     999990       93622.86
   10960.895 0.999990844727     999991      109226.67
   11354.111 0.999992370605     999993      131072.00
   11862.015 0.999993133545     999994      145635.56
   11862.015 0.999993896484     999994      163840.00
   12296.191 0.999994659424     999995      187245.71
   13418.495 0.999995422363     999996      218453.33
   13467.647 0.999996185303     999997      262144.00
   13467.647 0.999996566772     999997      291271.11
   13467.647 0.999996948242     999997      327680.00
   14319.615 0.999997329712     999998      374491.43
   14319.615 0.999997711182     999998      436906.67
   15736.831 0.999998092651     999999      524288.00
   15736.831 0.999998283386     999999      582542.22
   15736.831 0.999998474121     999999      655360.00
   15736.831 0.999998664856     999999      748982.86
   15736.831 0.999998855591     999999      873813.33
   17203.199 0.999999046326    1000000     1048576.00
   17203.199 1.000000000000    1000000
#[Mean    =       80.623, StdDeviation   =      242.151]
#[Max     =    17203.199, Total count    =      1000000]
#[Buckets =           24, SubBuckets     =         2048]
```

But it seems that my example has huge latency (it was run locally), can You help me with this benchmark? What can I enhance?

mjpt777 commented:
Before reviewing this in any detail you should note that it does not pass the build which should be a minimum.

Can I ask why does a simple latency test require fork-join pool? This seems total overkill.

segabriel commented:
@mjpt777 Hi
>Before reviewing this in any detail you should note that it does not pass the build which should be a minimum.

yeah, there are some problems with checkstyle.

> Can I ask why does a simple latency test require fork-join pool? This seems total overkill.

without this, I receive `Aeron exception occurred: io.aeron.exceptions.AeronException: reentrant calls not permitted during callbacks`

mjpt777 commented:
This still fails the build. I'm made few comments to help give some ideas where to start.

The approaches illustrates a number of fundamental misunderstandings about developing low-latency or high-throughput systems. To do much more work on this feels like free consultancy. Hopefully you have a few pointers to get started with. Next step would be learning to use a profiler and understanding the costs of your code. If you are seriously about learning how to do this better then maybe consider getting some professional coaching or training.

segabriel commented on aeron-samples/src/main/java/io/aeron/samples/mdc/AeronClientPing.java:
```diff
@@ -0,0 +1,205 @@
+            for (int i = 0; i < NUMBER_OF_MESSAGES; )
+            {
+
+                ExpandableArrayBuffer buffer = new ExpandableArrayBuffer(Long.BYTES);
```
Ok, but I didn't find a DirectBuffer allocator which I can use to allocate buffers inside it and reuse them to reduce allocate memory. 
I'd like to send buffers and receive them asynchronously even though it will in one thread of eventLoop. 
Ok, maybe I need to create only one buffer and reuse it. But I don't know about its potential size.
I used to have netty-buffers and I can allocate and release them easily.

mjpt777 commented:
It should be pretty clear from the results of the sample and your code that the approaches give very different performance results. It is best if you try to migrate your coding style to the samples than going the other way if you wish to retain performance. If you want guidance on how to build something more significant than the sample then look at how the Aeron Archive or Cluster is designed and implemented.

segabriel commented:
@mjpt777 Thanks a lot!

I rewrote my tests with already existing ping/pong examples. And I changed only URI channels which they should use and all examples were with enabled of embeddedMediaDriver (https://github.com/segabriel/aeron/pull/2). 
So results of existing ping/pong gave me: 

```c
#[Mean    =      145.634, StdDeviation   =       82.711]
#[Max     =    21200.895, Total count    =      1000000]
#[Buckets =           24, SubBuckets     =         2048]
```

and the same but with using of control endpoint gave me:

```c
#[Mean    =      200.594, StdDeviation   =      118.529]
#[Max     =    20479.999, Total count    =      1000000]
#[Buckets =           24, SubBuckets     =         2048]
```

and results of sending without waiting of its responses (because the previous tests sent one request and waited for its response, so I changed a bit this behavior):

```c
#[Mean    =   114167.109, StdDeviation   =     8795.130]
#[Max     =   141688.831, Total count    =     10000000]
#[Buckets =           24, SubBuckets     =         2048]
```

segabriel commented:
Then I added simple backpressure support (I send several limited messages and when the limit is achieved I stop sending and wait for responses when they will equal to the specified limit and then again begin sending) and it looks like better and pretty well:

here is the results for `aeron.sample.request=8`:

```c
#[Mean    =      149.115, StdDeviation   =       63.623]
#[Max     =     6471.679, Total count    =      1000000]
#[Buckets =           24, SubBuckets     =         2048]
```

for `aeron.sample.request=16`:

```c
#[Mean    =      167.663, StdDeviation   =       88.051]
#[Max     =    10395.647, Total count    =      1000000]
#[Buckets =           24, SubBuckets     =         2048]
```

and for `aeron.sample.request=32`:

```c
#[Mean    =      169.890, StdDeviation   =      125.346]
#[Max     =    10387.455, Total count    =      1000000]
#[Buckets =           24, SubBuckets     =         2048]
```
